File: 380 Insert Delete GetRandom O(1)/Java/RandomizedSet.java
import java.util.Arrays;
import java.util.Random;

public class RandomizedSet {
    private int[] arr;
    private int capacity;
    private int size;

    public RandomizedSet() {
        capacity = 10;
        arr = new int[capacity];
        size = 0;
    }

    public RandomizedSet(int capacity) {
        this.capacity = capacity;
        arr = new int[this.capacity];
        size = 0;
    }

    public int[] getArr() {
        return arr;
    }

    public boolean insert(int val) {
        for (int i = 0; i < size; i++) {
            if (arr[i] == val) {
                return false;
            }
        }

        if (size == capacity) {
            capacity *= 2;
            arr = Arrays.copyOf(arr, capacity);
        }

        arr[size++] = val;
        return true;
    }

    public boolean remove(int val) {
        if (size == 0) {
            return false;
        }

        for (int i = 0; i < size; i++) {
            if (arr[i] == val) {
                arr[i] = arr[size - 1];
                size--;
                return true;
            }
        }
        return false;
    }

    public int getRandom() {
        Random random = new Random();
        int randomIdx = random.nextInt(size);
        return arr[randomIdx];
    }
}

File: 380 Insert Delete GetRandom O(1)/Python/RandomizedSet.py
import random
from typing import Dict, List, Self


class RandomizedSet:
    arr: List[int]
    idxMap: Dict[int, int]

    def __init__(self: Self):
        self.arr = []
        self.idxMap = {}

    def search(self, val: int) -> bool:
        return val in self.idxMap
    
    def insert(self, val: int) -> bool:
        if self.search(val):
            return False
    
        self.arr.append(val)
        self.idxMap[val] = len(self.arr) - 1
        return True

    def remove(self, val: int) -> bool:
        if not self.search(val):
            return False
        
        idx: int = self.idxMap[val]
        self.arr[idx] = self.arr[-1]
        self.idxMap[self.arr[idx]] = idx
        self.arr.pop()
        del self.idxMap[val]
        return True

    def getRandom(self) -> int:
        return random.choice(self.arr)